### PR TITLE
Add policy for setting up Grafana alerts

### PR DESCRIPTION
**Pull request by premun**
This is a result of a discussion that happened here: https://github.com/dotnet/core-eng/issues/15725

All feedback is appreciated, I added people who were part of the discussion

[example alert](https://dev.azure.com/dnceng/internal/_wiki/wikis/DNCEng%20Services%20Wiki/917/-Alerts-Android-devices-disconnected)

### CODE REVIEW

**Comment by premun** [comment]
@markwilkie @ilyas1974 @tkapin can you help me review this policy and merge it please?

File: Documentation/Policy/GrafanaAlertsPolicy.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+- For every Grafana alert there is a page with alert guide tagged as `[Alert]` in the [internal wiki](https://dev.azure.com/dnceng/internal/_wiki/wikis/DNCEng%20Services%20Wiki/884/Wiki-for-Grafana-Alerts) dedicated to this specific alert
+  - The page is marked as `[Alert]` and contains the the category tag `!!Category-Alerts` so that is is listed in the [Alerts category page](https://dev.azure.com/dnceng/internal/_wiki/wikis/DNCEng%20Services%20Wiki/925/Category-Alerts)
+  - The page contains information about the alert, why it fired, what's the impact/urgency, what actions to take and more. Please see a [more exhaustive list below](#recommended-content-of-guides-to-alerts).
+- The body of the alert contains a link to the appropriate wiki page
+  - It's much easier to maintain the wiki page than the alert's description
+  - If the alert fires repeatedly (uses one of the notification channels with periodic reminders), include this information in the alert's description too.  
+    Example:  
+    ![image](https://user-images.githubusercontent.com/7013027/169300812-574c0ee7-ecc9-4e25-b163-4e5b64d43cd0.png)
+- If you encounter an alert missing this information or
```
Comment (riarenas) [inline Documentation/Policy/GrafanaAlertsPolicy.md:27]: This sentence is either incomplete, or the goal is to induce panic if an alert is missing the information.

**Review by riarenas — COMMENTED** [review]
Always up for more actionable alerts. 

Do we want to start thinking of ways we're going to make sure the policy is being followed? 

Maybe some sort of consistency check that makes sure all alerts have a corresponding guide that could be a separate leg in the PR builds for the repos that hold dashboards?

**Comment by premun** [comment]
Thanks for the review!

> Do we want to start thinking of ways we're going to make sure the policy is being followed?
> 
> Maybe some sort of consistency check that makes sure all alerts have a corresponding guide that could be a separate leg in the PR builds for the repos that hold dashboards?

My thinking was something like @ilyas1974 could enforce it piece by piece as FR receives alerts and works through them slowly? I think a person who just resolved an alert is a good candidate to describe how they did it and what resources they used when doing that. It could also spread the work among the team in an amortized manner? :)

File: Documentation/Policy/GrafanaAlertsPolicy.md (hunk 1 above)
Comment (premun) [reply Documentation/Policy/GrafanaAlertsPolicy.md:27]: Ugh, who knows what I wanted to put there! I will think of something new 😅

**Comment by riarenas** [comment]
I think that makes sense for existing alerts, but I'd rather we had some way to ensure that new alerts are up to standards. We can say that it's up to PR reviewers to ensure that happens, but it's very easy for them to be missed. 

Once we've made our existing alerts better, I'd like to avoid us being in a cycle of:

* Alert gets checked in without a corresponding guide for x or y reason
* FR has to scramble to solve the alert the first time it happens
* Now FR person who suffered through it has to also remediate the missing guides.

for every new alert.

File: Documentation/Policy/GrafanaAlertsPolicy.md (hunk 1 above)
Comment (premun) [reply Documentation/Policy/GrafanaAlertsPolicy.md:27]:
```suggestion
```

**Comment by markwilkie** [comment]
Given that @ilyas1974 is diving into the "docs are a jungle" work, I'll let him take a look.

I love the fact that we'd have guidance.  I agree with Ricardo too that including how we plan to provide visibility to how we're doing per policy seems important.

**Comment by premun** [comment]
We could have a warning in the `helix-services` pipeline that would verify a link presence in the `message` attribute in the JSON?

File: Documentation/Policy/GrafanaAlertsPolicy.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+- Alerts are self-explanatory and approachable
+  - By approachable we mean that even non-SMEs should be able to get an idea on the next steps or at least have a point of contact to go to.
+- Person handling the alert understands the rough impact and urgency and is able to communicate this to our partners
+- Alert investigation and potential mitigation can happen faster as useful links/resources/documentation are available in alert's guide
+
+## Policy
+
+Whenever you're setting up a new alert in Grafana make sure that:
+
+- Panels in Grafana dashboards have a description that explains what the graph/chart/table/panel shows and why we measure that particular metric.
```
Comment (garath) [inline Documentation/Policy/GrafanaAlertsPolicy.md:16]: Consider also requiring panels include a Link to their alert wiki page.

File: Documentation/Policy/GrafanaAlertsPolicy.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+Some alerts have obvious next steps - go look at this chart, check the heartbeats table, query Kusto in a certain way. Other actions can be communication towards our partners when we know the impact. Common actions can get their own page and be linked from several alerts.
+
+**Usual causes/remediation**  
+Some problems just come back and this is a piece of information that is hard to get thanks to the way we do FR rotation since we don't have the contiunity insight. Feel free to mention this characteristic if we keep seeing the same thing and how it gets usually resolved. Mention whether it's transient or outside of our control completely.
+
+**Links to related resources**  
+These can be repos, specific code, Azure resources, 3rd party systems and other. This should speed up investigations for obvious reasons but also simply prevent mistakes such as searching in the wrong Azure subscription, rebooting the wrong SF cluster, navigating to a wrong storage account when looking for SQL tables..
+
+**List of SMEs**  
+Do you know that this whole area was implemented by Billy? Is Billy still around? Well.. add them so that they can help if things go bananas.
```
Comment (garath) [inline Documentation/Policy/GrafanaAlertsPolicy.md:46]: This is inspired.

File: Documentation/Policy/GrafanaAlertsPolicy.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+## Recommended content of guides to alerts
+
+The actual content of the guide will differ for every alert but here are some useful pointers it should include.
+
+**Alert explanation**  
+Short description of what we are measuring, when and why we alert and what it means for our systems (urgency, impact..). We can list affected components and describe symptoms which can help answer FR questions during outages.
+
+**Actions to take**  
+Some alerts have obvious next steps - go look at this chart, check the heartbeats table, query Kusto in a certain way. Other actions can be communication towards our partners when we know the impact. Common actions can get their own page and be linked from several alerts.
```
Comment (garath) [inline Documentation/Policy/GrafanaAlertsPolicy.md:37]: "Step by step instructions are great"

File: Documentation/Policy/GrafanaAlertsPolicy.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  - It's much easier to maintain the wiki page than the alert's description
+  - If the alert fires repeatedly (uses one of the notification channels with periodic reminders), include this information in the alert's description too.  
+    Example:  
+    ![image](https://user-images.githubusercontent.com/7013027/169300812-574c0ee7-ecc9-4e25-b163-4e5b64d43cd0.png)
+
+
+## Recommended content of guides to alerts
+
+The actual content of the guide will differ for every alert but here are some useful pointers it should include.
+
```
Comment (garath) [inline Documentation/Policy/GrafanaAlertsPolicy.md:32]: I think it's also good to have a brief description of the service or area of service that the alert is involved in. The context can be really useful for debugging.

**Comment by garath** [comment]
Maybe we put 👍 / 👎 feedback buttons on alerts? (Only kinda kidding.)

File: Documentation/Policy/GrafanaAlertsPolicy.md (hunk 2 above)
Comment (premun) [reply Documentation/Policy/GrafanaAlertsPolicy.md:16]: Yeah, this is a good point

Comment (premun) [reply Documentation/Policy/GrafanaAlertsPolicy.md:16]: I tried to put a markdown style link in the description and it worked! Cool! Added

Comment (garath) [reply Documentation/Policy/GrafanaAlertsPolicy.md:16]: It could also be a "panel link".

image

**Review by ChadNedzlek — APPROVED** [review]

**Comment by ChadNedzlek** [comment]
A warning would get ignored (our builds have had hundreds of warnings for ages... every now and then I go clean them up, but in general, no one looks at things that aren't failing).

I'm sort of fine with this being a code review/after the fact when an alert triggers situation.  Honestly, for most alerts, until they trigger the first time, it's _hard_ to write that documentation, because you don't necessarily know what investigation is helpful.  And if it never actually triggers (we have a lot of _really_ conservative alerts that basically boil down to "the sky is falling"), spending a bunch of time to write up a bunch of documentation isn't a huge win (in terms of time spent or the complexity of our documentation).  Also, the mere _presence_ of the stuff isn't really enough... it's easy to write useless comments/docs, so a code reviewer really should make sure that the documentation is understandable to someone other than the person who wrote it... so there is already a burden on reviewers.

File: Documentation/Policy/GrafanaAlertsPolicy.md (hunk 2 above)
Comment (premun) [reply Documentation/Policy/GrafanaAlertsPolicy.md:16]: Ah, will fix this then

**Comment by premun** [comment]
> A warning would get ignored (our builds have had hundreds of warnings for ages... every now and then I go clean them up, but in general, no one looks at things that aren't failing).

Correct me if I'm wrong but OSOB and helix-services have 0 warnings and we tend to them as they have a certain function always. At least in OSOB every warning does, I wasn't able to find one in helix-services? Or are they silenced there?

> Also, the mere presence of the stuff isn't really enough... it's easy to write useless comments/docs, so a code reviewer really should make sure that the documentation is understandable to someone other than the person who wrote it... so there is already a burden on reviewers.

I agree that a review for alert docs is important and it's hard to get them right but a mere presence of links leading to the right place (e.g. links to the correct AppInsights with logs) would go a long way from where we are now. The way we name in things in cloud is sometimes quite unfortunate and since I personally don't work with our services often, to this day I have still failed to memorize many of the things so each alert investigation starts by me trying to find the right resource.

Whatever we agree on, I would like to lower the bar for any kind of doc improvement as much as possible (this is where I am trying to get from where we are with this whole effort).

**Comment by ChadNedzlek** [comment]
arcade-services and helix-services only don't have warnings because I rather regularly go through and clean them up (usually when I'm doing other sweeping changes, and they are making it hard for me to tell what I might have broken). Anything that doesn't fail a PR, especially in arcade services (where you likely won't even look at the build if it passes, because it will just be a green dot in GitHub) really doesn't affect behavior much, so generally isn't worth writing/maintaining.

I like the idea of using the wiki for most of the actual documention, and leaving the alerts/panels pretty low weight. Editting the wiki is basically zero-barrier.  Editing the panel/alert and then importing that into the dashboard, and then making a PR, just to add a little extra context is a HUGE burden, and likely why no one bothers. :-)  I wish the AzDO wiki was really a wiki, and you could have dead links that lead to a "create this page" thing... then we could just automatically generate the links for every alert.  BUT, AzDO wikis are, in fact, not wikis in that way, so we can't. (GitHub wikis were, so it was awesome when we had one of those).

**Comment by premun** [comment]
@garath @riarenas I feel like we're kind of following this policy even without it being checked in. However, when we define a new alert, it might be helpful to point people somewhere. Maybe we can just merge this and be done with it?

**Comment by garath** [comment]
I'm fine with this. @ilyas1974 should review as well as he has been doing work in normalizing the format of FR-ish wiki things, including alerts.

**Review by garath — APPROVED** [review]

**Review by MattGal — APPROVED** [review]

**Comment by premun** [comment]
Oops, sorry, didn't notice it was Matt but Ilya with the second approval

**Pull request merged**